nucleus/payara-modules/healthcheck-core/src/main/java/fish/payara/nucleus/healthcheck/preliminary/HoggingThreadsHealthCheck.java

```diff
@@ -91,65 +142,107 @@ public String getDescription() {
+                            "Thread", thread, //
+                            "Usage%", String.valueOf(percentage), //
+                            "Threshold%", String.valueOf(threshold), //
+                            "Method", initialMethod, //
+                            "Exited", String.valueOf(!initialMethod.equals(getMethod(info))));
+                    hoggingThreadCount.incrementAndGet();
+                    hoggingThreadMaxDuration.updateAndGet(value -> Math.max(value, totalTimeHogging));
+                });
+        collector
+        .collect("HoggingThreadCount", hoggingThreadCount)
+        .collect("HoggingThreadDuration", hoggingThreadMaxDuration);
+    }
 
-            if (c == -1 || u == -1)
-                continue;
+    @Override
+    public void collect(MonitoringWatchCollector collector) {
+        if (options == null || !options.isEnabled() || !supported) {
+            return;
+        }
+        collector.watch("ns:health HoggingThreadCount", "Hogging Threads", "count")
+            .amber(0, -2, false, null, null, false)
+            .red(1, -2, false, null, null, false);
+    }
 
-            ThreadTimes times = threadTimes.get(id);
-            if (times == null) {
-                times = new ThreadTimes();
-                times.setId(id);
-                times.setName(threadInfo.getThreadName());
-                times.setStartCpuTime(c);
-                times.setEndCpuTime(c);
-                times.setStartUserTime(u);
-                times.setEndUserTime(u);
-                threadTimes.put(id, times);
+    private void acceptHoggingThreads(Map<Long, ThreadCpuTimeRecord> recordsById, HoggingThreadConsumer consumer) {
+        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
+        long now = System.currentTimeMillis();
+        long currentThreadId = Thread.currentThread().getId();
+        int retryCount = options.getRetryCount();
+        int threshold = options.getThresholdPercentage().intValue();
+        for (long threadId : bean.getAllThreadIds()) {
+            if (threadId == currentThreadId)
+                continue;
+            final long cpuTimeInNanos = bean.getThreadCpuTime(threadId);
+            if (cpuTimeInNanos == -1)
+                continue;
+            long cpuTime = TimeUnit.NANOSECONDS.toMillis(cpuTimeInNanos); 
+            // from here all times are in millis
+            ThreadCpuTimeRecord record = recordsById.get(threadId);
+            if (record == null) {
+                record = new ThreadCpuTimeRecord();
+                recordsById.put(threadId, record);
             } else {
-                times.setStartCpuTime(times.getEndCpuTime());
-                times.setStartUserTime(times.getEndUserTime());
-                times.setEndCpuTime(c);
-                times.setEndUserTime(u);
-
-                long checkTime = getOptions().getUnit().toMillis(getOptions().getTime());
```

Comment:
NB. using the interval from the options is only an approximation of the actual time passed in the measurement interval. As this now can be different intervals for the check run by the health check service and the monitoring data collection I changed this to use the actual time passed. This is also more accurate.